core/src/Kokkos_LogicalMemorySpace.hpp

```diff
@@ -0,0 +1,395 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_LOGICALMEMORYSPACE_HPP
+#define KOKKOS_LOGICALMEMORYSPACE_HPP
+
+#include <cstring>
+#include <string>
+#include <iosfwd>
+#include <typeinfo>
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_Concepts.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+
+#include <impl/Kokkos_Traits.hpp>
+#include <impl/Kokkos_Error.hpp>
+#include <impl/Kokkos_SharedAlloc.hpp>
+
+#include "impl/Kokkos_HostSpace_deepcopy.hpp"
+#include "impl/Kokkos_Profiling.hpp"
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+
+/// \class LogicalMemorySpace
+/// \brief
+///
+/// LogicalMemorySpace is a space that is identical to another space, but
+/// differentiable by name and template argument
+
+template <class Namer, class BaseSpace, class DefaultExecutionSpace = void,
+          bool SharesAccessWithBase = true>
+class LogicalMemorySpace {
+ public:
+  //! Tag this class as a kokkos memory space
+  typedef LogicalMemorySpace<Namer, BaseSpace, DefaultExecutionSpace,
+                             SharesAccessWithBase>
+      memory_space;
+  typedef size_t size_type;
+
+  /// \typedef execution_space
+  /// \brief Default execution space for this memory space.
+  ///
+  /// Every memory space has a default execution space.  This is
+  /// useful for things like initializing a View (which happens in
+  /// parallel using the View's default execution space).
+
+  using execution_space =
+      typename std::conditional<std::is_void<DefaultExecutionSpace>::value,
+                                typename BaseSpace::execution_space,
+                                DefaultExecutionSpace>::type;
+
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+
+  /**\brief  Default memory space instance */
+  LogicalMemorySpace() : underlying_allocator(){};
```

Comment:
Why is that one not defaulted?

Reply:
Because I didn't think to, good comment, I'll do that